pkg/events/redis/store.go

```diff
@@ -298,49 +302,53 @@ func (ps *PubSubStore) SubscribeWithHistory(
 
 	matchNames := xMessageHasEventName(names...)
 
+	// We expect that the subscriber can ingest `tail` events at a time.
+	eventCountLimit := int64(tail)
+	switch {
+	case eventCountLimit < 8:
+		eventCountLimit = 8
+	case eventCountLimit > 1024:
+		eventCountLimit = 1024
+	}
+
 	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
-			args := &redis.XReadArgs{
-				Count: 10,
-				Block: time.Second,
+		args := &redis.XReadArgs{
+			Count: eventCountLimit,
+			Block: random.Jitter(8*time.Second, 0.2),
+		}
+		for _, s := range state {
```

Comment:
The diff here looks bigger than it is. I have removed the outer `select` (since `XRead` will return `ctx.Err()` on `ctx.Done()`), but the only changes are to `args.Count` and `args.Block`.